C3 dashboard charts

vojtechszocs opened the pull request:
C3Charting supports following charts:

![miq-c3-bar](https://cloud.githubusercontent.com/assets/648971/12397410/62d47082-be0d-11e5-8c06-c694d51d9b64.png)

![miq-c3-bar-stacked](https://cloud.githubusercontent.com/assets/648971/12397427/6ef8603a-be0d-11e5-911c-dac73bb50576.png)

![miq-c3-column](https://cloud.githubusercontent.com/assets/648971/12397435/75142710-be0d-11e5-837b-14797583866f.png)

![miq-c3-column-stacked](https://cloud.githubusercontent.com/assets/648971/12397438/78ee07ac-be0d-11e5-84e6-e39562134da8.png)

![miq-c3-donut](https://cloud.githubusercontent.com/assets/648971/12397439/7c1febfc-be0d-11e5-8a27-340fb5cdef78.png)

![miq-c3-pie](https://cloud.githubusercontent.com/assets/648971/12397444/7e82f1e6-be0d-11e5-8187-e810c5ac19c9.png)

Modify C3Charting.priority to activate.

Fryguy commented:
@martinpovolny Please review.

martinpovolny commented on app/helpers/charting_helper.rb:
```diff
@@ -1,5 +1,6 @@
 module ChartingHelper
   def chart_remote(a_controller, options)
+    p :chart_remote
```
`p`

DNNX commented on app/helpers/charting_helper.rb:
```diff
@@ -15,24 +16,35 @@ def chart_remote(a_controller, options)
     elsif Charting.backend == :jqplot
       jqplot_sample(options.slice(:id, :bgcolor, :width, :height))
+    elsif Charting.backend == :c3
+      c3chart_sample
```
Is it possible to use `case Charting.backend when ...` there? Also in `chart_remote` and `chart_local`.

DNNX commented on lib/charting/c3_charting.rb:
```diff
@@ -0,0 +1,81 @@
+  end
+
+  def data_ok?(data)
+    !!YAML.load(data) && !YAML.load(data)[:options]
```
`YAML.load` is not an extremely fast operation, so it is probably better to evaluate it once.

jrafanie replied:
:+1: Yes, we should not YAML.load twice here.

himdel replied:
Should fix the Jqplot version as well then, it should have the same problem

DNNX commented on lib/report_formatter/c3_series.rb:
```diff
@@ -0,0 +1,23 @@
+    end
+
+    def sum
+      reduce(0.0) { |sum, datum| sum + datum[:value].to_f }
```
```ruby
def sum
  super { |datum| datum[:value].to_f }
end
```

DNNX commented on lib/report_formatter/c3_series.rb:
```diff
@@ -0,0 +1,23 @@
+module ReportFormatter
+  class C3Series < Array
+    def initialize(type = :flat)
```
Is `type` unused?

Fryguy commented:
@vojtechszocs Please see the rubocop comments [above](https://github.com/ManageIQ/manageiq/pull/6223#issuecomment-172601912) as well.

martinpovolny commented on app/assets/javascripts/application.js:
```diff
@@ -51,8 +51,8 @@
-//= require d3/d3
-//= require c3/c3
+//= require c3
+//= require d3
```
ok, according to @himdel, this is actually better than the original but should not change anything

abonas replied:
> ok, according to @himdel, this is actually better than the original but should not change anything meaningful

why is it better?

himdel replied:
Because the intent is to include the whole library, require c3 clearly means that. require c3/c3 means include a particular file from the library, even though it's the only one there so the end result is exactly the same.

abonas replied:
> Because the intent is to include the whole library, require c3 clearly means that. require c3/c3 means include a particular file from the library, even though it's the only one there so the end result is exactly the same.

:+1:

martinpovolny replied:
https://github.com/rails/sprockets part "Index files are proxies for folders". This way seems to be consistent with the docs.

martinpovolny replied:
oh, too late for the party ;-)

abonas commented:
@vojtechszocs is the donut going to have the same look and feel like the pf donut? 
https://www.patternfly.org/patterns/donut-chart/
container dashboards are using the pf donut.
cc @zeari

martinpovolny commented:
@abonas : the look and feel is not in the scope of this PR. We want the UX people to do the look and feel for us. I guess that will deal with what you are concerned about.

The goal of the PR by @vojtechszocs is to have the charts rendered by C3 with the correct data in all places.

vojtechszocs commented:
Hi guys, I've updated my initial commit (replaced the original one) and I think I've addressed your review comments. I'm not a Ruby expert (actually a total beginner) so please bear with me.

vojtechszocs commented:
> @vojtechszocs is the donut going to have the same look and feel like the pf donut? 

As @martinpovolny mentioned the styling of C3 charts is not in scope here. This is the first phase where we aim to have C3 charts working with actual data.

PatternFly does not support all types of C3 charts used in this PR yet, see `c3ChartDefaults` in [patternfly.js](https://github.com/patternfly/patternfly/blob/master/dist/js/patternfly.js). So even though we could use default PF donut chart styling (`getDefaultDonut`), we're still missing PF styling for column, bar and pie charts.

For now, this PR uses custom C3 chart styles, see `app/assets/javascripts/miq_c3_config.js`. Once PatternFly supports all types of C3 charts used in ManageIQ dashboard, just editing this file should do the trick, hopefully.

martinpovolny commented:
@vojtechszocs : do you have any updates? Or is this up-to-date at this point?

vojtechszocs commented:
@martinpovolny I've tried to address all review comments in commit vojtechszocs@4b5cbb5 and I only see two more RuboCop warnings in `lib/charting/c3_charting.rb` so I can fix them.

Besides that, I'm done with my changes, unless you have any more comments. Since `C3Charting.priority < JqplotCharting.priority` the existing jqPlot based charting backend should still be in effect by default and we can improve C3 charting backend as needed.

martinpovolny commented:
Ok @vojtechszocs , can you rebase it then? Let me do a detailed review and test and let's get this merged.

martinpovolny commented:
@vojtechszocs : running you PR, I can see charts in reports, report editor, full-screen report. But no charts in widgets or full-screen from widget.
Is that expected? Or is it a problem on my side?

martinpovolny commented:
this renders:
```
{"miqChart":"Bar","data":{"columns":[["linux_centos_1",5.0,1.0,0,0,0,0,0,0,0],["linux_generic_2",178.0,116.0,31.0,23.0,10.0,0,0,0,0],["linux_redhat_3",1223.0,352.0,300.0,246.0,231.0,3.0,0,1.0,0],["unknown_4",196.0,0,0,1.0,0,0,2.0,0,0],["windows_generic_5",41.0,0,0,0,2.0,0,0,0,0],["linux_suse_6",16.0,0,0,0,0,0,0,0,0],["linux_ubuntu_7",2.0,0,0,0,0,0,0,0,0]],"names":
{"linux_centos_1":"linux_centos","linux_generic_2":"linux_generic","linux_redhat_3":"linux_redhat","unknown_4":"unknown","windows_generic_5":"windows_generic","linux_suse_6":"linux_suse","linux_ubuntu_7":"linux_ubuntu"}},"axis":{"x":{"categories":["no value","cfme...","cfme...","cfme...","cfme...","qebl...","qebl...","ibm-...","Other"]}}}
```

this does not:
````
{"miqChart":"ParallelThreedColumn","data":{"columns":[["OpenStack / Other: 1",0],["RedHat / Other: 11",11],["VMware / Other: 44",44]]}}
```

with
```
Uncaught Error: url or json or rows or columns is required.c3_chart_internal_fn.init @ c3.self-ca27322e3461cbb9c6df194e6e11fe6b8aa183236f7c0191e420ee55c6df9243.js?body=1:92Chart @ c3.self-ca27322e3461cbb9c6df194e6e11fe6b8aa183236f7c0191e420ee55c6df9243.js?body=1:34c3.generate @ c3.self-ca27322e3461cbb9c6df194e6e11fe6b8aa183236f7c0191e420ee55c6df9243.js?body=1:58(anonymous function) @ report_only?rr_id=10000000000001&type=hybrid:184fire @ jquery.self-977e28a4e7fded7698789e394a585c6339d54c0ad1537f498a40d2800098a521.js?body=1:3183self.fireWith @ jquery.self-977e28a4e7fded7698789e394a585c6339d54c0ad1537f498a40d2800098a521.js?body=1:3313done @ jquery.self-977e28a4e7fded7698789e394a585c6339d54c0ad1537f498a40d2800098a521.js?body=1:8755(anonymous function) @ jquery.self-977e28a4e7fded7698789e394a585c6339d54c0ad1537f498a40d2800098a521.js?body=1:9121
Navigated to http://localhost:3000/dashboard/report_only?rr_id=10000000000001&type=hybrid
```

martinpovolny commented:
Oh now I see, you wrote:

```
    C3Charting supports following charts:
    
    - Bar + StackedBar
    - Column + StackedColumn
    - Donut
    - Pie
```

martinpovolny commented:
We can move on. E.g. mapping `PieTheeed` to `Pie` etc.

@vojtechszocs : please, do rebase, I'll do my best to get this merged and then create issues for the missing features that we can attack one-by-one.

chessbyte commented:
@vojtechszocs pretty charts!

vojtechszocs commented:
Whoops, sorry everyone, bad rebase! I'll rebase again.

vojtechszocs commented:
> @vojtechszocs : please, do rebase, I'll do my best to get this merged and then create issues for the missing features that we can attack one-by-one.

Thanks! Indeed, this PR just covers some basic chart types, we can fill the gaps via follow-up PRs as needed.

himdel commented:
@martinpovolny https://www.patternfly.org/patterns/bar-chart/
